File: Reservoir-Computing-in-Julia/softmax_test.py
import numpy as np

y_true = np.array([[1, 0, 0], [0, 0, 1], [1, 0, 0]])
y_pred = np.array([[0.7, 0.2, 0.1], [0.1, 0.1, 0.8], [0.3, 0.3, 0.4]])

loss = -1/len(y_true) * np.sum(np.sum(y_true * np.log(y_pred)))
print(loss)

import torch
import torch.nn as nn

y_true = torch.LongTensor([0,2,0])
y_pred = torch.tensor([[0.7, 0.2, 0.1], [0.1, 0.1, 0.8], [0.3, 0.3, 0.4]], dtype=torch.float32)
ce_loss = nn.CrossEntropyLoss()

loss = ce_loss(y_pred, y_true)
print(loss.item())

File: Reservoir-Computing-in-Julia/softmax_test.jl
using NNlib
# testing softmax function
y_true = [1 0 0; 0 0 1; 1 0 0]
y_pred = [0.7 0.2 0.1; 0.1 0.1 0.8; 0.3 0.3 0.4]
println(softmax(y_true))

loss_value = -1/size(y_true, 1) * sum(y_true .* log.(y_pred))

function my_softmax(x)
    return exp.(x) ./ sum(exp.(x), dims=1)
end

my_softmax(y_true)